File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/benchmark/benchmark.rb
# iterative solution
def find_shortest_string(arr)
  arr.reduce do |shortest, string|
    string.length < shortest.length ? string : shortest
  end
end

# recursive solution
def find_shortest_string_recursive(arr)
  # base case
  return arr.first if arr.length == 1
  # recursive call
  result = find_shortest_string_recursive(arr[1..-1])
  # return the shortest
  arr.first.length <= result.length ? arr.first : result
end

def benchmark
  start_time = Time.now

  1000.times do 
    yield
  end

  (Time.now - start_time) / 1000
end

puts "Iterative Solution: #{benchmark { find_shortest_string(['lily', 'sunflower', '', 'dahlia', 'lilac']) }}"
puts "Recursive Solution: #{benchmark { find_shortest_string_recursive(['lily', 'sunflower', '', 'dahlia', 'lilac']) }}"

File: 02-week-2--recursion/04-day-4--recursive-find-shortest-string/benchmark/benchmark.js
// Iterative Solution:
function findShortestString(arr) {
  return arr.reduce((shortest, string) => 
    string.length < shortest.length ? string : shortest
  );
}

// Recursive Solution:
function findShortestStringRecursive(arr) {
  // base case
  if(arr.length === 1) {
    return arr[0];
  }
  // recursive call
  const result = findShortestStringRecursive(arr.slice(1));
  return arr[0].length <= result.length ? arr[0] : result;
}

function benchmark(callback) {
  const startTime = Date.now();

  for (let i = 0; i < 1000; ++i) {
    callback();
  }

  return (Date.now() - startTime) / 1000;
}

console.log('Iterative Solution:', benchmark(() => findShortestString(['lily', 'sunflower', '', 'dahlia', 'lilac'])));
console.log('Recursive Solution:', benchmark(() => findShortestStringRecursive(['lily', 'sunflower', '', 'dahlia', 'lilac'])));
